compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/AggregatedOpInterfaceImpl.cpp

```diff
@@ -175,6 +175,49 @@ static Value computeMatmul(OpBuilder &builder, Location loc, AffineMap lhsMap,
   return genericOp.getResult(0);
 }
 
+static Value applyMask(OpBuilder &builder, Location loc, AffineMap qkMap,
+                       AffineMap maskMap, Value qk, Value mask) {
+
+  SmallVector<AffineMap> compressedMaps =
+      compressUnusedDims(SmallVector<AffineMap>{qkMap, maskMap});
+  qkMap = compressedMaps[0];
+  maskMap = compressedMaps[1];
+
+  SmallVector<utils::IteratorType> iteratorTypes(qkMap.getNumDims(),
+                                                 utils::IteratorType::parallel);
+
+  Value zero = builder.create<arith::ConstantOp>(
+      loc, builder.getFloatAttr(getElementTypeOrSelf(qk.getType()), 0.0));
+  Value negInf = builder.create<arith::ConstantOp>(
+      loc, builder.getFloatAttr(getElementTypeOrSelf(qk.getType()), -INFINITY));
+
+  auto genericOp = builder.create<linalg::GenericOp>(
+      loc, qk.getType(), SmallVector<Value>{mask}, qk,
+      SmallVector<AffineMap>{maskMap, qkMap}, iteratorTypes,
+      [&](OpBuilder &b, Location loc, ValueRange args) {
+        Value qkVal = args[1];
+        Value maskVal = args[0];
+
+        // TODO: Replace bool mask condition once treated as i1 (instead of i8)
+        if (maskVal.getType().isInteger()) {
+          maskVal =
+              b.create<arith::TruncIOp>(loc, builder.getI1Type(), maskVal);
+          maskVal = b.create<arith::SelectOp>(loc, maskVal, zero, negInf);
+        } else {
+          maskVal = convertScalarToDtype(b, loc, maskVal, qkVal.getType(),
+                                         /*isUnsignedCast=*/false);
+          // Scaling to compensate for base-2 softmax
+          Value log2e = b.create<arith::ConstantOp>(
+              loc, b.getFloatAttr(qkVal.getType(), M_LOG2E));
+          maskVal = b.create<arith::MulFOp>(loc, maskVal, log2e);
```

Comment:
Because the softmax operation on qk^T + mask is in base 2, we must include this additional log2e scaling in order to make the softmax act as a base e operation (as torch does)

Reply:
Ah and also we have to scale mask by log2e before applying it too. We put it here to do both scaling and mask application in the same generic instead of two separate ones